### ISSUE TITLE
Should the colNames and colModel be updated on the DOM on column sort in jQgrid?

### ISSUE THREAD

**Issue by reypm** [free-jqgrid/jqGrid#346]
I am having a weird issue and I haven't a clue in how to fix this. 

Let's say I have a grid which is build on the server side - yes, we've a class `DynamicGrid` (PHP) receiving a few parameters and building all the HTML/JS needed by jQgrid to render a grid.

We're also storing the grid setup in a MongoDB as for example: columns order and some other custom stuff.

When the grid loads for first time the data is displayed properly.

![screenshot-localhost-2017-06-21-05-53-09](https://user-images.githubusercontent.com/627886/27378510-1ebafa9c-5646-11e7-95e4-65e8cb6266fb.png)

And this is the JSON I got from the server side:

```json
{
  "page": "1",
  "total": 3,
  "records": 135,
  "rows": [
    {
      "id": "1",
      "cell": [
        "1|1|1|1|1|\/exportfile\/gridexport\/key\/63ca0eeef3b172204368b879e93ebd896d6ef8f7",
        "1",
        "23096",
        "Wonderware West - Houston",
        "0",
        "100494-2",
        "CoorsTek Inc",
        "open",
        "2017-02-20",
        "2018-02-19",
        "1444",
        "current",
        "AutoQuote",
        "16",
        "9120",
        "$1,789.93",
        "1",
        "Wonderware",
        "10-7001R",
        "$37,588.53",
        "Standard",
        "Percentage Support",
        "EndUser",
        "600 9th St.",
        "Golden",
        "Colorado",
        "80401",
        "UNITED STATES",
        "NORTH AMERICA",
        "North America",
        "79",
        "SABMiller plc"
      ]
    },
    ...
  ]
}
```
I can even sort the columns and it still working (column AgreementId has been reordered):

![screenshot-localhost-2017-06-21-05-56-13](https://user-images.githubusercontent.com/627886/27378594-675e4c4a-5646-11e7-964f-5eef73814c58.png)

And this time the JSON I got from the server side looks as:

```json
{
  "page": "1",
  "total": 3,
  "records": 135,
  "rows": [
    {
      "id": "1",
      "cell": [
        "1|1|1|1|1|\/exportfile\/gridexport\/key\/01045853abc74286cc886d6d1effc9d26630c4f0",
        "1",
        "100494-2",
        "0",
        "23096",
        "Wonderware West - Houston",
        "CoorsTek Inc",
        "open",
        "2017-02-20",
        "2018-02-19",
        "1444",
        "current",
        "AutoQuote",
        "16",
        "9120",
        "$1,789.93",
        "1",
        "Wonderware",
        "10-7001R",
        "$37,588.53",
        "Standard",
        "Percentage Support",
        "EndUser",
        "600 9th St.",
        "Golden",
        "Colorado",
        "80401",
        "UNITED STATES",
        "NORTH AMERICA",
        "North America",
        "79",
        "SABMiller plc"
      ]
    }
    ...
  ]
}
```

As soon as I paginate the grid the values are misaligned and won't display properly anymore.

![screenshot-localhost-2017-06-21-05-59-04](https://user-images.githubusercontent.com/627886/27378763-dfeaa280-5646-11e7-816a-e891e2690450.png)

(compare against previous values)

This is the JSON I got when paginate the grid:

```json
{
  "page": "2",
  "total": 3,
  "records": 135,
  "rows": [
    {
      "id": "307",
      "cell": [
        "307|1|1|1|1|\/exportfile\/gridexport\/key\/63ca0eeef3b172204368b879e93ebd896d6ef8f7",
        "307",
        "100632-2",
        "0",
        "41383",
        "Wonderware West - Houston",
        "Frito-Lay Inc.",
        "open",
        "2018-02-01",
        "2019-01-31",
        "29616",
        "current",
        "AutoQuote",
        "16",
        "26",
        "$1,793.66",
        "1",
        "Wonderware",
        "10-7002R",
        "$34,079.45",
        "Premium",
        "Percentage Support",
        "EndUser",
        "2800 Silver Star Road",
        "Orlando",
        "Florida",
        "32808",
        "UNITED STATES",
        "NORTH AMERICA",
        "North America",
        "68",
        "Pepsi"
      ]
    }
    ...
  ]
}
```
Why the data isn't displayed properly on the grid? What I am doing wrong or what I am missing here?

> Note: I am using the latest version of this free-jQgrid. There is a question open [here](https://stackoverflow.com/questions/44657495/should-the-colnames-and-colmodel-be-updated-on-the-dom-on-column-sort-in-jqgrid?) in SO with a similar example (different data though) when I am running into the same issue

**Comment by OlegKi** [comment]
Could you create the demo, which reproduces the problem? I can help you only if I can reproduces and debug the the code. You can create demo in [jsfiddle](https://jsfiddle.net/) for example. You can use https://jsfiddle.net/OlegKi/m7rqjyqo/ as the starting point and modify it to your application. If you need to load the data from the server, then you can use Echo service of jsfiddle. See https://jsfiddle.net/OlegKi/92mhochf/ as an example.

**Comment by reypm** [comment]
Hi @OlegKi I've tried my best [here](https://jsfiddle.net/reynierpm/u5hj4xxm/2/) but I can't find why is not working properly. I am not sure even in how to provide you with the data on each pagination could you help me out?

**Comment by OlegKi** [comment]
Your original code uses `datatype: "json"` and your current code a little dirty and it works only in case of specific format of input data (array of strings). Thus I'd recommend you to use Echo service: https://jsfiddle.net/OlegKi/u5hj4xxm/3/. It's display the data now, but what is the problem, which you need to solve?

In general, the usage of array format of `"cell":` (see input data) is mostly bad idea. It's not good to use `name` of `colModel` with values which contains `"."` symbol. It works, but you can have some problems later maintain the application. You will be much more flexible if you would use data items like
```
{
    "quote_id": "79|1|1|1|1|\/exportfile\/gridexport\/key\/63ca0eeef3b172204368b879e93ebd896d6ef8f7",
    "QuoteID": "79",
    "ParentAgreementNumber": "100536-7",
    ...
}
```
instead of
```
{
    "id": "79",
    "cell": ["79|1|1|1|1|\/exportfile\/gridexport\/key\/63ca0eeef3b172204368b879e93ebd896d6ef8f7", "79", "100536-7", "0", "49926", "Wonderware France", "Procter & Gamble", "open", "2009-09-01", "2012-08-31", "58343", "current", "AutoQuote", "89", "117", "$5,833.72", "1", "Wonderware", "10-7001R", "$86,975.22", "Standard", "Percentage Support", "EndUser", "Rue Andre Durouchez", "Amiens Cedex 2", "Not Applicable", "80013", "FRANCE", "EURA", "EMEA", "73", "Procter & Gamble"]
  }
```
You can use additionally `key: true` property in the column `QuoteID`, you can use `label` property in `colModel` and remove `colNames` option. After changing the format of input data you will be able remove all hidden columns from `colModel` and to use `additionalProperties` parameter instead (`additionalProperties: ["agreement_id", "quote_type", ..., "AccountName"]`). It improves performance of the grid.

**Comment by reypm** [comment]
@OlegKi awesoooommmmeeeee many thanks man, now an issue has been fixed. The issue I was having was related to the data not following it's column on pagination it was not showing properly (see the examples at the images on this post). Now changing the JSON as you said make it to work like a charm.

Can you show me an example of this: _After changing the format of input data you will be able remove all hidden columns from colModel and to use additionalProperties parameter instead (additionalProperties: ["agreement_id", "quote_type", ..., "AccountName"]). It improves performance of the grid._ I am lost with this one

**Comment by OlegKi** [comment]
You are welcome!

You can modify the last demo (https://jsfiddle.net/OlegKi/u5hj4xxm/3/) to the code/data, which you use now and I will show you how one can optimize it more.

**Comment by reypm** [comment]
@OlegKi done, [here](https://jsfiddle.net/reynierpm/ajf50cuf/1/) I've to add the `label` as suggested working on it at the meantime

**Comment by OlegKi** [comment]
What is your current problem? I can gives you some additional common advice, but I'm not sure that it's what you need.

For example, I don't recommend you to use comma separated names in `colModel` because in some cases such names will be interpreted as path to the property. Additionally point will be used in CSS selectors, which require escaping it in all selectors.

I made some changes in the demo. See https://jsfiddle.net/OlegKi/ajf50cuf/3/
I extended information saved to `colModel` (the properties `table` and `field`). Now you are free in choosing the `name`. I used `jsonmap` to read the data from the currently formatted input data and modified `beforeSubmitCell` to use `table` and `field`:
```JavaScript
beforeSubmitCell: function(rowid, cellname, value, iRow, iCol) {
      var cm = $(this).jqGrid("getGridParam", "colModel")[iCol];
      //var tableField = cellname.split(".");
      return {
        table: cm.table, //tableField[0],
        field: cm.field, // tableField[1],
        fieldvalue: value
      };
}
```
By the way the usage of `jsonmap` allow to read the data in **two formats**: the format, which you currently use and in the object format like
```JavaScript
{
    "actions": "1|1|1|1|1|\/exportfile\/gridexport\/key\/01045853abc74286cc886d6d1effc9d26630c4f0",
    "quote": {
        "QuoteID": "100494-12",
        "StartDate": "2017-02-22",
        "EndDate": "2018-03-21",
        "quote_status": "open"
    },
    "distributor": {
    	"DistributorName": "Wonderware West - Houston",
    	"Region": "North America"
    },
    "agreement": {
        "ParentAgreementNumber": "0",
        "AgreementNumber": "23096"
    },
...
```
See https://jsfiddle.net/OlegKi/ajf50cuf/4/ which contains modified format only in the first item. You can see that the item will be displayed in the grid in the same way like all other items, formatted in the old format.

**Comment by reypm** [comment]
@OlegKi something weird is going on on my side and I don't know why or how to reproduce it. The problem is with the formatter applied to `actions`. If I use `options.rowId` then I got a weird string append to the ID itself as for example `jqg+Id` where `Id` is a number which makes the icon to fail since it's not a valid ID in the other hand if I use `data[0]` then it works properly, any ideas?

Also is there any way to block the Action columns and by `block` I mean: not allow to sort it, not allow to order, not allow to be added or removed. Is this possible?

Again, thx for your time

**Comment by OlegKi** [comment]
The reason: you need to specify how jqGrid should get rowid from the input data. One way would be to add `key: true` property in `"quote.quote_id"` column. Better way: to use the option `jsonReader: { id: "quote.quote_id"}` to inform jqGrid about the name of property, which contains the rowid. After including the option one can remove unneeded hidden column `"quote.quote_id"`. See https://jsfiddle.net/OlegKi/ajf50cuf/6/

**Comment by reypm** [comment]
that will not work for me since this grid is not tied to only such data is generated dynamically so the key could change on each case.

Any idea on the second comment on the paragraph above?

**Comment by OlegKi** [comment]
I'n not sure what you mean. `data[0]` or any other part of formatter `GridQuoteActions` has no relation to the **rowid**. Your previously formatter data had `id` property (see `{id:123, cell:[...]}`). If you want to use `options.rowId` in the code of the formatter and if you want that the rowid will be sent during cell editing then you should specify rowid in some way.

For example, you can just add `id` property to the input data together to all other properties (line you did before). One more way: you can place `key: true` to the column of `colModel`, which contains rowid information. It's good only if you really want to include the rowid information in `colModel`. The third option is the usage of `jsonReader.id`. You should decide yourself, which way is the best one in your application.

**Comment by reypm** [comment]
@OlegKi I am using `key: true` on the colModel but somehow is not working for me. 

Any ideas regarding this one:

> Also is there any way to block the Action columns and by block I mean: not allow to sort it, not allow to order, not allow to be added or removed. Is this possible?

**Comment by OlegKi** [comment]
If you provide the demo, which uses `key: true` in one column , which hold the required information, and which is not work then I could help you. You can comment `jsonReader: { id: "quote.quote_id"},` line in my last demo and uncomment the definition of `"quote.quote_id"`. After that you can verify with Developer Tools that the values of `id` attribute of rows (`<tr>` element) still correctly set. See https://jsfiddle.net/OlegKi/ajf50cuf/7/.

To "block" the column `"actions"` I would recommend you to add the same properties, which uses `template: "actions"` (see [the lines](https://github.com/free-jqgrid/jqGrid/blob/v4.14.1/js/jquery.fmatter.js#L213-L221)):
```JavaScript
autoResizable: false, frozen: true, fixed: true, hidedlg: true,
resizable: false, sortable: false, search: false, editable: false, viewable: false
```

**Comment by reypm** [comment]
@OlegKi I am having another issue which is not happening in the sample [here](https://jsfiddle.net/reynierpm/ajf50cuf/9/) as you can see there I can't sort any column not even actions. Now if I try the same code on my side I can order all of them. Don't know what the difference is. However the issue still on my side, I can order the column Actions and I can perform actions on it, why is that?

**Comment by OlegKi** [comment]
I'm not sure what you mean under "I can't sort any column". If you click on the column header of any column then the column will be correctly sorted.

**Comment by reypm** [comment]
Ups sorry typo|wrong word, is "I can't order any column" you can change the column position at all. (this happen on the example) at my side I can order all of them "Actions" included, the JS looks exactly as the one on the example.

**Comment by OlegKi** [comment]
Please could you express more clear what you mean. What you mean "you can change the column position"? You can change the order of items in `colModel` for example. If you mean changing the column order by drag & drop of the column headers, then one have to include `sortable: true` option (the demo use it) and to include jQuery UI sortable. The demo uses `https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/themes/redmond/jquery-ui.min.css` already. If I add `https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.11.4/jquery-ui.min.js` in the list of "External resources" then we get https://jsfiddle.net/OlegKi/ajf50cuf/10/, where one can reorder the column by drag and drop. 

By the way I see that you set `"autoResizable": false` and `"resizable": false` for all column. It prevent the user from reading full text in the column header and the user can't resize the columns. [The wiki article](https://github.com/free-jqgrid/jqGrid/wiki/Auto-adjustment-of-the-width-of-column) describes autoresizing feature, which allows to resize the column by *content* by double-click in the resize area (between column headers). I personally use `"resizable": false` only in columns, which contains fixed size elements like icons or buttons.

It seems to me that all the last discussion is far from your original question, which you posted in the issue. It's probably better to close the issue and to post new one if you have new problem.

**Comment by reypm** [comment]
Closing this issue ....
